# HG changeset patch
# User Bobby Johnson <rjohnson19@gmail.com>
# Date 1600028397 14400
#      Sun Sep 13 16:19:57 2020 -0400
# Node ID b3ed3176733e7a4a45bdf75ebd698f8b5a0f45c0
# Parent  7fc438bde1953d75e0d9c471456240fe3dcb8ed4
Restore ability to show UnifiedComplete results on empty searches.

If topsites is disabled in general, or for the urlbar,
the topsites provider should be considered inactive, rather than
not showing any results. Undoes parts of Bug 1623666.

diff --git a/browser/components/urlbar/UrlbarProviderTopSites.jsm b/browser/components/urlbar/UrlbarProviderTopSites.jsm
--- a/browser/components/urlbar/UrlbarProviderTopSites.jsm
+++ b/browser/components/urlbar/UrlbarProviderTopSites.jsm
@@ -65,17 +65,26 @@ class ProviderTopSites extends UrlbarPro
   /**
    * Whether this provider should be invoked for the given context.
    * If this method returns false, the providers manager won't start a query
    * with this provider, to save on resources.
    * @param {UrlbarQueryContext} queryContext The query context object
    * @returns {boolean} Whether this provider should be invoked for the search.
    */
   isActive(queryContext) {
-    return !queryContext.searchString;
+    // We'll show Top Sites when there is no search string,
+    // Top Sites are enabled in general, and enabled for the URL Bar.
+    return (
+      !queryContext.searchString &&
+      Services.prefs.getBoolPref(
+        "browser.newtabpage.activity-stream.feeds.system.topsites",
+        false
+      ) &&
+      UrlbarPrefs.get("suggest.topsites")
+    );
   }
 
   /**
    * Gets the provider's priority.
    * @param {UrlbarQueryContext} queryContext The query context object
    * @returns {number} The provider's priority for the given query.
    */
   getPriority(queryContext) {
@@ -86,34 +95,16 @@ class ProviderTopSites extends UrlbarPro
    * Starts querying.
    * @param {UrlbarQueryContext} queryContext The query context object
    * @param {function} addCallback Callback invoked by the provider to add a new
    *        result. A UrlbarResult should be passed to it.
    * @note Extended classes should return a Promise resolved when the provider
    *       is done searching AND returning results.
    */
   async startQuery(queryContext, addCallback) {
-    // If system.topsites is disabled, we would get stale or empty Top Sites
-    // data. We check this condition here instead of in isActive because we
-    // still want this provider to be restricting even if this is not true. If
-    // it wasn't restricting, we would show the results from UnifiedComplete's
-    // empty search behaviour. We aren't interested in those since they are very
-    // similar to Top Sites and thus might be confusing, especially since users
-    // can configure Top Sites but cannot configure the default empty search
-    // results. See bug 1623666.
-    if (
-      !UrlbarPrefs.get("suggest.topsites") ||
-      !Services.prefs.getBoolPref(
-        "browser.newtabpage.activity-stream.feeds.system.topsites",
-        false
-      )
-    ) {
-      return;
-    }
-
     let sites = AboutNewTab.getTopSites();
 
     let instance = {};
     this.queries.set(queryContext, instance);
 
     // Filter out empty values. Site is empty when there's a gap between tiles
     // on about:newtab.
     sites = sites.filter(site => site);
